pkg/libartifact/store/store.go

```diff
@@ -254,6 +256,147 @@ func (as ArtifactStore) Add(ctx context.Context, dest string, paths []string, op
+			len(options.Digest) == 0 && len(options.Title) == 0 {
+			return fmt.Errorf("the artifact consists of several blobs and the target %q is not a directory and neither digest or title was specified to only copy a single blob", target)
+		}
+
+		var digest digest.Digest
+		if len(arty.Manifest.Layers) == 1 {
+			digest = arty.Manifest.Layers[0].Digest
+		} else {
+			digest, _, err = findDigestAndTitle(arty, options)
+			if err != nil {
+				return err
+			}
+		}
+
+		return copyImageBlobToFile(ctx, imgSrc, digest, target)
+	}
+
+	if len(options.Digest) > 0 || len(options.Title) > 0 {
+		digest, title, err := findDigestAndTitle(arty, options)
+		if err != nil {
+			return err
+		}
+		filename := generateArtifactName(title, digest)
+		return copyImageBlobToFile(ctx, imgSrc, digest, filepath.Join(target, filename))
+	}
+
+	for _, l := range arty.Manifest.Layers {
+		title := l.Annotations[specV1.AnnotationTitle]
+		filename := generateArtifactName(title, l.Digest)
+		err := copyImageBlobToFile(ctx, imgSrc, l.Digest, filepath.Join(target, filename))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func generateArtifactName(title string, digest digest.Digest) string {
+	filename := title
+	if len(filename) == 0 {
+		// No filename given, use the digest. But because ":" is not a valid path char
+		// on all platforms replace it with "-".
+		filename = strings.ReplaceAll(digest.String(), ":", "-")
+	}
+	return filename
+}
+
+func findDigestAndTitle(arty *libartifact.Artifact, options *libartTypes.ExtractOptions) (digest.Digest, string, error) {
+	var (
+		digest digest.Digest
+		title  string
+	)
+	for _, l := range arty.Manifest.Layers {
+		if options.Digest == l.Digest.String() {
+			if len(digest.String()) > 0 {
+				return digest, "", fmt.Errorf("more than one match for the digest %q", options.Digest)
+			}
+			digest = l.Digest
+			title = l.Annotations[specV1.AnnotationTitle]
+		}
+		if len(options.Title) > 0 {
+			if val, ok := l.Annotations[specV1.AnnotationTitle]; ok &&
+				val == options.Title {
+				if len(digest.String()) > 0 {
+					return digest, "", fmt.Errorf("more than one match for the title %q", options.Title)
+				}
+				digest = l.Digest
+				title = val
+			}
+		}
+	}
+	if len(digest.String()) == 0 {
+		if len(options.Title) > 0 {
+			return digest, "", fmt.Errorf("no blob with the title %q", options.Title)
+		}
+		return digest, "", fmt.Errorf("no blob with the digest %q", options.Digest)
+	}
+	return digest, title, nil
+}
+
+func copyImageBlobToFile(ctx context.Context, imgSrc types.ImageSource, digest digest.Digest, target string) error {
+	src, _, err := imgSrc.GetBlob(ctx, types.BlobInfo{Digest: digest}, nil)
+	if err != nil {
+		return fmt.Errorf("failed to get artifact file: %w", err)
+	}
+	defer src.Close()
+	dest, err := os.Create(target)
+	if err != nil {
+		return fmt.Errorf("failed to create target file: %w", err)
+	}
+	defer dest.Close()
+
+	// TODO use reflink is possible
```

Comment:
is this a long term todo, or something you would do as part of this PR?

Reply:
This is easy to do, but as usual I don't want to duplicate code and the c/image function is internal so I have to export it first then vendor again, etc... So not something I plan to do for this PR but I will be working on that soon enough.